### PR TITLE
Enabled logging analyzer on new repo

### PR DESCRIPTION
**Pull request by maryamariyan**
The dotnet/extensions code: https://github.com/dotnet/extensions/tree/master/src/Logging/Logging.Analyzers

cc: @mavasani @sharwell @pgovind

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:53]: @mavasani for `CA1727` and `CA1848` I want them disabled by default. Is it sufficient to just use `RuleLevel.Disabled` as illustrated here? Or we also need to use `isEnabledByDefaultInAggressiveMode`?

Follow up question: is there a way we can test the disabled diagnostics in the disabled mode? Asking because regardless of how I set them up in the tests they are always triggered.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:55]: This doesn't seem correct to me.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:200]: There shouldn't be a placeholder here.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:135]: Use `invocation.ReportDiagnostic`

**Comment by maryamariyan** [comment]
Was caught up with conflicts and checking to see if the CI build is successful before pushing up a WIP commit trying to address using operation analysis instead of syntax analysis.

Once that is done, I will address the remaining feedback.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:131]: This reports a diagnostic on every usage of any method in this type. Why not mark this type obsolete in dotnet/runtime instead?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:231]: Can't this part also be done using IOperation?

**Comment by maryamariyan** [comment]
the MemberData test cases for `CA2255IsProducedForFormatArgumentCountMismatch` fail on the latest update. Looking into it.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx
Comment (maryamariyan) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:200]: I found some existing cases where placeholders are used in the resx in this repo.

```
> git grep -c "{0}" -- *.resx

src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx:39
src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/BannedApiAnalyzerResources.resx:3
src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:134
src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:149
src/NetAnalyzers/Core/Microsoft.NetFramework.Analyzers/MicrosoftNetFrameworkAnalyzersResources.resx:25
src/PerformanceSensitiveAnalyzers/CSharp/Analyzers/AnalyzersResources.resx:2
src/PublicApiAnalyzers/Core/Analyzers/PublicApiAnalyzerResources.resx:18
src/Roslyn.Diagnostics.Analyzers/Core/RoslynDiagnosticsAnalyzersResources.resx:13
src/Text.Analyzers/Core/TextAnalyzersResources.resx:18
tools/FileIssues/Resources.resx:13
```

Comment (Youssef1313) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:200]: @maryamariyan They can be used only for messages, not descriptions.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: @Youssef1313 this line is out of date. Is it recommended to not use Diagnostic.Create?

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: you're suggesting to remove `CA1848` altogether? I'm not opposed to that

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: Yes I believe it's recommended to use the extension methods.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:219]:
```suggestion
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:128]:
```suggestion
            if (containingType.Equals(loggerExtensionsType, SymbolEqualityComparer.Default))
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:134]:
```suggestion
            else if (!containingType.Equals(loggerType, SymbolEqualityComparer.Default) && !containingType.Equals(loggerMessageType, SymbolEqualityComparer.Default))
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:146]:
```suggestion
                if (containingType.Equals(loggerMessageType, SymbolEqualityComparer.Default))
```

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:147]:
```suggestion
                if (containingType.Equals(loggerMessageType, SymbolEqualityComparer.Default))
```

**Review by Youssef1313 — COMMENTED** [review]
Since there is no language specific logic now. I believe this can be moved up from `CSharp` to `Core` without any extra work other than unit tests.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:51]: I'm not sure about this.

**Comment by maryamariyan** [comment]
The test failures still need looking into. @sharwell and I plan on looking at this again today.

**Comment by Youssef1313** [comment]
@maryamariyan I'm not seeing test failures in CI. The failure in CI is due to auto-generated files being outdated. Can you run `msbuild /t:pack /v:m` in the repository root? This should update the following files:

-  `src\NetAnalyzers\Microsoft.CodeAnalysis.NetAnalyzers.md`
-  `src\NetAnalyzers\Microsoft.CodeAnalysis.NetAnalyzers.sarif`

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: Yes that's what I'm suggesting (but this requires obsoleting so a warning is still issued).

**Comment by maryamariyan** [comment]
@Youssef1313 the reason the CI is green is because the test cases failing for `CA2255IsProducedForFormatArgumentCountMismatch` are commented out.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineTests.cs
Comment (maryamariyan) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineTests.cs:42]: here.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: @pakrym I chatted with @shirhatti as well and he mentioned he would be fine with skipping diagnostic `CA1848`. Since you were the original author, I wanted to make sure you are not against this either.

Comment (pakrym) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: Sounds good to me.

Comment (pakrym) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: Maybe we can turn it into a code-fix one day.

File: src/NetAnalyzers/CSharp/AnalyzerReleases.Unshipped.md
Comment (mavasani) [inline src/NetAnalyzers/CSharp/AnalyzerReleases.Unshipped.md:3]: Lets avoid adding new rules to `CSharp/AnalyzerReleases.Unshipped.md`. Can you instead create a base type with the diagnostic descriptors in the [core](https://github.com/dotnet/roslyn-analyzers/tree/main/src/NetAnalyzers/Core) language-agnostic NetAnalyzers project? This way all the shipped/unshipped rules get added to the same files: `src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md` and `src/NetAnalyzers/Core/AnalyzerReleases.Shipped.md`

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (mavasani) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:89]: You should not use `RuleLevel.BuildWarning`. It is reserved for highly critical analyzers that indicate serious functional bug in code, something that is guaranteed to make the code behave badly at runtime. These are considered build breaking rules as /warnaserror would cause anyone updating to new .NET SDK to break their builds, something we never want for naming or usage or performance rules

File: src/NetAnalyzers/CSharp/AnalyzerReleases.Unshipped.md
Comment (buyaa-n) [inline src/NetAnalyzers/CSharp/AnalyzerReleases.Unshipped.md:8]: @maryamariyan please update the proposed Severity for these 2

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: the API review approved the fifth diagnostic as warning since its usage would result in a runtime exception

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (maryamariyan) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs]: these two were approved as hidden in the API review meeting

**Comment by maryamariyan** [comment]
The PR is ready for review :)

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:181]: Please rereview and update messaging as mentioned in the API review, for example these could be something like this:
```suggestion
  <data name="LoggerMessageDiagnosticNumericsInFormatStringTitle" xml:space="preserve">
    <value>Named placeholders are preferred than numeric in logging format string</value>
  </data>
  <data name="LoggerMessageDiagnosticNumericsInFormatStringMessage" xml:space="preserve">
    <value>Use named placeholders instead numeric placeholders</value>
  </data>
  <data name="LoggerMessageDiagnosticNumericsInFormatStringDescription" xml:space="preserve">
    <value>Using named placeholders in logging format string is more descriptive than numeric placeholders.</value>
```

File: src/NetAnalyzers/RulesMissingDocumentation.md
Comment (buyaa-n) [inline src/NetAnalyzers/RulesMissingDocumentation.md:11]: Mistakenly removed?

**Comment by jmarolf** [comment]
@maryamariyan if this is going to be enabled by default in the .NET 6 SDK I suggest we add a performance test for it here: https://github.com/dotnet/roslyn-analyzers/tree/main/src/PerformanceTests/Tests/Enabled this can happen in a follow up PR in the RC1/RC2 timeframe

**Comment by maryamariyan** [comment]
> @maryamariyan have we run this against the dotnet/runtime and dotnet/aspnetcore repos? Would want to identify any false positives there.

I am testing this against the aspnetcore repo to make sure there are no issues

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:126]: is the comment intended?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:288]:
```suggestion
        private static bool FindLogParameters(IMethodSymbol methodSymbol,[NotNullWhen(true)] out IParameterSymbol? message, out IParameterSymbol? arguments)
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:166]: is the TODO done?

**Comment by buyaa-n** [comment]
> I am testing this against the aspnetcore repo to make sure there are no issues

Overall LGTM, if the testing with aspnetcore gives false positives for warning/info level analyzers please fix them before merging

**Review by buyaa-n — APPROVED** [review]

**Comment by maryamariyan** [comment]
for future reference here is a sample set exercising all 5 diagnostics assuming all are enabled:

image

with 

image

File: src/NetAnalyzers/RulesMissingDocumentation.md
Comment (maryamariyan) [reply src/NetAnalyzers/RulesMissingDocumentation.md:11]: no, the 

> build.cmd -pack 

script removes it

Comment (maryamariyan) [reply src/NetAnalyzers/RulesMissingDocumentation.md:11]: @jmarolf calling out to check if it is ok to delete this line. Seems reasonable since the docs page is not missing anymore:

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2251

**Comment by maryamariyan** [comment]
> @maryamariyan if this is going to be enabled by default in the .NET 6 SDK I suggest we add a performance test for it here: https://github.com/dotnet/roslyn-analyzers/tree/main/src/PerformanceTests/Tests/Enabled this can happen in a follow up PR in the RC1/RC2 timeframe

OK, will address in a follow up PR

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:130]: This is passing two arguments to `CreateDiagnostic`, while the message of this rule has only one placeholder.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:315]:
```suggestion
                }
                else if (parameter.IsParams &&
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineTests.cs:193]:
```suggestion
                TestCode = code,
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:302]: Is there a missing parenthesis here or the current form is intended?
`&&` has higher precedence. So a name matching `messageFormat` or `formatString` will meet the condition even if it's not string.

**Comment by buyaa-n** [comment]
> for future reference here is a sample set exercising all 5 diagnostics assuming all are enabled:

The message structure looks good to me, but the `Named holes` sounds weird, not sure that is the term used for C# cc @terrajobst

**Comment by maryamariyan** [comment]
> The message structure looks good to me, but the Named holes sounds weird, not sure that is the term used for C# cc @terrajobst

I renamed to "named placeholder" from "named hole", since that works as well

**Comment by terrajobst** [comment]
> I renamed to "named placeholder" from "named hole", since that works as well

💯

**Pull request merged**
